solr/core/src/java/org/apache/solr/highlight/UnifiedSolrHighlighter.java

```diff
@@ -431,14 +435,33 @@ protected Locale parseLocale(String language, String country, String variant) {
 
     @Override
     protected Predicate<String> getFieldMatcher(String field) {
-      // TODO define hl.queryFieldPattern as a more advanced alternative to hl.requireFieldMatch.
 
       // note that the UH at Lucene level default to effectively "true"
       if (params.getFieldBool(field, HighlightParams.FIELD_MATCH, false)) {
         return field::equals; // requireFieldMatch
-      } else {
-        return NOT_REQUIRED_FIELD_MATCH_PREDICATE;
       }
+
+      String[] queryFieldPattern =
+          params.getFieldParams(field, HighlightParams.QUERY_FIELD_PATTERN);
+      if (queryFieldPattern != null && queryFieldPattern.length != 0) {
+
+        Collection<String> indexedFields = solrIndexSearcher.getDocFetcher().getIndexedFieldNames();
+
+        Predicate<String> predicate = null;
+        for (String f : expandWildcardsInHighlightFields(indexedFields, queryFieldPattern)) {
+          if (predicate == null) {
+            predicate = f::equals;
+          } else {
+            predicate = predicate.or(f::equals);
+          }
```

Comment:
this results in a O(N) predicate.  I think it would be more efficient to simply build up a HashSet and have a simple predicate to check it -- O(1).

Reply:
Good point, thanks! Will do.

Reply:
Changed in https://github.com/apache/solr/pull/757/commits/00e3ccdd9485c6e3627d95dc0c0321ff248a0766 commit.